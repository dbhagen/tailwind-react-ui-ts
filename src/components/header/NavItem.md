Individual menu items in navigation bar.
