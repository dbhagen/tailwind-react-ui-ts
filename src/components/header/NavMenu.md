Navigation bar used in `Header`.
