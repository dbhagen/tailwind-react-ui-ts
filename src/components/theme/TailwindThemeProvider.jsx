import React from 'react'
import PropTypes from 'prop-types'
import merge from 'lodash.merge'

import defaultTheme from './defaultTheme'
import TailwindTheme from './TailwindTheme'

const TailwindThemeProvider = ({ theme, children }) => {
  const mergedTheme = merge({}, defaultTheme, theme)

  return (
    <TailwindTheme.Provider value={mergedTheme}>
      {children}
    </TailwindTheme.Provider>
  )
}

TailwindThemeProvider.propTypes = {
  theme: PropTypes.shape({}),
  children: PropTypes.node,
}

TailwindThemeProvider.defaultProps = {
  theme: {},
  children: undefined,
}

export default TailwindThemeProvider
