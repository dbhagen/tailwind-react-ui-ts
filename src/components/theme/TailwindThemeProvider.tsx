import React from 'react'
import PropTypes from 'prop-types'
import merge from 'lodash.merge'

import defaultTheme from './defaultTheme'
import TailwindTheme from './TailwindTheme'

// TODO: Remove Any Types
const TailwindThemeProvider = ({ theme, children }: { theme: any, children: any }) => {
  const mergedTheme = merge({}, defaultTheme, theme)

  return (
    <TailwindTheme.Provider value={mergedTheme}>
      {children}
    </TailwindTheme.Provider>
  )
}

TailwindThemeProvider.propTypes = {
  theme: PropTypes.shape({}),
  children: PropTypes.node,
}

TailwindThemeProvider.defaultProps = {
  theme: {},
  children: undefined,
}

export default TailwindThemeProvider
