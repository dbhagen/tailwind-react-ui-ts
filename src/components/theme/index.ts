export { default as defaultTheme } from './defaultTheme'
export { default as TailwindTheme } from './TailwindTheme'
export { default as TailwindThemeProvider } from './TailwindThemeProvider'
export { default as withTheme } from './withTheme'
export { default as useThemeValue } from './useThemeValue'
