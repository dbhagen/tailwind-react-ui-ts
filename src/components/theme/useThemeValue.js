import { tailwindPropToClassName } from '../tailwind'

export default (prefix, value, userClassNames = '') => {
  if (userClassNames.includes(prefix)) {
    return false
  }

  return tailwindPropToClassName(prefix, value)
}
