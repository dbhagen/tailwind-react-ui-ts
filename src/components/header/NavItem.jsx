import React from 'react'
import PropTypes from 'prop-types'

import { withTheme } from '../theme'
import { Touchable } from '../primitives'

const NavItem = ({
  theme,
  is,
  children,
  header: { style, screen },
  active,
  ...rest
}) => {
  const textColor = style.text || theme.textColors.on.primary
  const bgColor = style.bg || theme.brandColors.primary

  const responsive = screen
    ? {
      [`m-${screen}`]: { t: 0, r: theme.spacing.sm }
    }
    : {}

  return (
    <Touchable
      is={is}
      focusable
      text={!active ? style.text : style.bg}
      bg={active ? textColor : undefined}
      bg-hocus={textColor}
      text-hocus={bgColor}
      p={{ x: theme.spacing.md, y: theme.spacing.sm }}
      m={{ t: theme.spacing.sm }}
      rounded={theme.radius}
      block
      aria-current={active ? 'page' : undefined}
      {...responsive}
      {...rest}
    >
      {children}
    </Touchable>
  )
}

NavItem.propTypes = {
  theme: PropTypes.shape({}).isRequired,
  is: PropTypes.oneOfType([PropTypes.string, PropTypes.func, PropTypes.object]),
  children: PropTypes.node,
  header: PropTypes.shape({
    style: PropTypes.object,
    screen: PropTypes.oneOfType([PropTypes.bool, PropTypes.string])
  }),
  active: PropTypes.bool
}

NavItem.defaultProps = {
  is: 'a',
  children: undefined,
  header: { style: {}, screen: 'lg' },
  active: false
}

export { NavItem as component }
export default withTheme(NavItem)
