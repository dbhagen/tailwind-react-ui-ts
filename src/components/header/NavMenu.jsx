import React from 'react'
import PropTypes from 'prop-types'

import { withTheme } from '../theme'
import { Box } from '../primitives'
import { Title } from '../typography'
import { withTransition } from '../utils'
import { List } from '../list'

import NavItem from './NavItem'

const NavItemWrapper = props => <li role="none" {...props} />

const NavMenu = ({
  theme,
  transition,
  is,
  children,
  header,
  list,
  ...rest
}) => {
  const transitionStyles = {
    entering: { maxHeight: '0', position: 'absolute' },
    entered: { maxHeight: '100vh' },
  }
  const headingId = `${header.id}-menu`

  const responsive = {
    nav: header.screen
      ? {
          [`w-${header.screen}`]: 'auto',
          [`flex-${header.screen}`]: true,
        }
      : {},
    menu: header.screen
      ? {
          [`flex-${header.screen}`]: true,
          [`m-${header.screen}`]: { b: 0 },
        }
      : {},
  }

  return (
    <Box
      is={is}
      overflow="hidden"
      w="full"
      flex="grow"
      items="center"
      h={!header.collapsable ? 12 : undefined}
      style={
        header.collapsable
          ? {
              transition: 'max-height 500ms',
              maxHeight: '0',
              ...transitionStyles[transition],
            }
          : undefined
      }
      id={`${header.id}-nav`}
      aria-labelledby={headingId}
      aria-expanded={header.collapsable ? header.open : undefined}
      role="navigation"
      {...responsive.nav}
      {...rest}
    >
      <Title level={2} id={headingId} visuallyHidden>
        Site menu
      </Title>
      <List
        reset
        flex="grow"
        role="menu"
        listItemIs={NavItemWrapper}
        {...responsive.menu}
        {...list}
      >
        {React.Children.map(
          children,
          child =>
            child.type === NavItem &&
            React.cloneElement(child, { header, role: 'menuitem' }),
        )}
      </List>
      {React.Children.map(children, child => child.type !== NavItem && child)}
    </Box>
  )
}

NavMenu.propTypes = {
  theme: PropTypes.shape({}).isRequired,
  transition: PropTypes.string,
  is: PropTypes.oneOfType([PropTypes.string, PropTypes.func, PropTypes.object]),
  children: PropTypes.node,
  header: PropTypes.shape({
    open: PropTypes.bool,
    collapsable: PropTypes.bool,
    screen: PropTypes.oneOfType([PropTypes.bool, PropTypes.string]),
  }),
  list: PropTypes.shape({}),
}

NavMenu.defaultProps = {
  is: 'nav',
  children: undefined,
  header: {
    open: false,
    collapsable: false,
    screen: 'lg',
  },
  transition: 'entering',
  list: {},
}

export { NavMenu as component }
export default withTheme(withTransition(NavMenu, { inState: 'header.open' }))
