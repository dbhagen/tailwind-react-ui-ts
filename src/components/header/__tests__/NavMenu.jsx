import React from 'react'
import { shallow } from 'enzyme'

import { component as NavMenu } from '../NavMenu'
import NavItem from '../NavItem'
import { defaultTheme } from '../../theme'

const setup = (testProps = {}) => {
  const props = Object.assign(
    {
      header: { open: true, collapsable: true },
      transition: 'entered',
      theme: defaultTheme,
    },
    testProps,
  )

  const wrapper = shallow(
    <NavMenu {...props}>
      <NavItem>About</NavItem>
      <NavItem>Blog</NavItem>
      <button type="button">Download</button>
    </NavMenu>,
  )

  return {
    props,
    wrapper,
  }
}

describe('NavMenu', () => {
  it('renders matching snapshot', () => {
    const { wrapper } = setup()

    expect(wrapper).toMatchSnapshot()
  })
})
