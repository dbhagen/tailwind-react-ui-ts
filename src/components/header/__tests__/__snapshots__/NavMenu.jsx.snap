// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavMenu renders matching snapshot 1`] = `
<Box
  aria-expanded={true}
  aria-labelledby="undefined-menu"
  flex="grow"
  id="undefined-nav"
  inline={false}
  inlineBlock={false}
  is="nav"
  items="center"
  overflow="hidden"
  role="navigation"
  style={
    Object {
      "maxHeight": "100vh",
      "transition": "max-height 500ms",
    }
  }
  w="full"
>
  <WithTheme(Title)
    id="undefined-menu"
    level={2}
    visuallyHidden={true}
  >
    Site menu
  </WithTheme(Title)>
  <WithTheme(List)
    flex="grow"
    listItemIs={[Function]}
    reset={true}
    role="menu"
  >
    <WithTheme(NavItem)
      header={
        Object {
          "collapsable": true,
          "open": true,
        }
      }
      key=".0"
      role="menuitem"
    >
      About
    </WithTheme(NavItem)>
    <WithTheme(NavItem)
      header={
        Object {
          "collapsable": true,
          "open": true,
        }
      }
      key=".1"
      role="menuitem"
    >
      Blog
    </WithTheme(NavItem)>
  </WithTheme(List)>
  <button
    key=".2"
    type="button"
  >
    Download
  </button>
</Box>
`;
